### PR TITLE
199: Resource packetfabric_billing_modify_order

### LINKED ISSUE #199: resource packetfabric_billing_modify_order
**Billing** (missing from DID 1.10)

Missing [Upgrade an existing Connection Speed, Term, or Type](https://docs.packetfabric.com/api/v2/swagger/#/Billing/modify_order) for: 

- Backbone VC
- Point-to-Point
- Hosted Cloud (AWS, Azure, Google, IBM, Oracle)
- Dedicated Cloud (AWS, Azure, Google)
- Cloud Router (AWS, Azure, Google, IBM, Oracle)

New resource `packetfabric_billing_modify_order`

**Details:**

POST [/v2/billing/services/{circuit_id}/modify](https://docs.packetfabric.com/api/v2/swagger/#/Billing/modify_order)
⚠️📈 You can only increase speed or term (not decrease) and only service from metro to long-haul
ℹ️ update the existing resources and handle the change of supported attributes

- Backbone VC: can only update subscription_term and/or speed
resource ``packetfabric_backbone_virtual_circuit``

- Point-to-Point: can only update subscription_term
resource ``packetfabric_point_to_point``

- Hosted Cloud (AWS, Azure, Google, IBM, Oracle): can only update speed
resource ``packetfabric_cs_<aws/azure/google/oracle/ibm>_hosted_connection``

- Dedicated Cloud (AWS, Azure, Google): can only update only subscription_term and/or service_class
resource ``packetfabric_cs_<aws/azure/google>_dedicated_connection``

- Cloud Router (AWS, Azure, Google, Oracle, IBM, IPsec, Port): can only update speed
resource ``packetfabric_cloud_router_connection_<aws/azure/google/oracle/ibm/ipsec/port>``

Check if connection can be upgraded checking [/v2/services/cloud-routers/{circuit_id}/connections/{connection_circuit_id}/upgrade/options](https://api-beta.dev.packetfabric.net/openapi/index.html#/Cloud%20Router%20Connections/cloud_routers_connection_upgrade_options)

Checklist:
- [ ] tested locally
- [ ] added/updated acceptance tests
- [ ] add examples (in example folder)
- [ ] generate docs

### PR DESCRIPTION
**Pull request by pontovinte**
Signed-off-by: Yrineu Rodrigues <yrineu@pontovinte.com>

## Description

Resource packetfabric_billing_modify_order
Closes #199 

## Checklist

- [x] tested locally
- [ ] added/updated acceptance tests
- [x] add examples (in example folder)
- [x] generate docs

### CODE REVIEW

File: internal/provider/resource_billing.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+				Description:  "New speed of the circuit.",
+			},
+			"billing_product_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+				Description:  "Billing type for VCs. Options are longhaul_dedicated, longhaul_usage, longhaul_hourly, surcharge, metro_dedicated.",
```
Comment (rjouhann) [inline internal/provider/resource_billing.go:52]: @pontovinte actually here probably better to do:

`ValidateFunc: validation.StringInSlice([]string{"longhaul_dedicated", "longhaul_usage", "surcharge", "metro_dedicated"}, true),`

And remove the options from the description, right?

**Review by elXaptation — APPROVED** [review]

**Comment by medzin** [comment]
Could this feature be implemented in the resources that created these orders (e.g. `packetfabric_point_to_point`) so the user will not have to modify the data of one resource via a second resource? If it is impossible to implement, then we should probably throw a custom error when the user changes the subscription terms in these resources e.g. "Unable to modify subscription terms in packetfabric_point_to_point please use packetfabric_billing_modify_order instead!".

**Comment by rjouhann** [comment]
> Could this feature be implemented in the resources that created these orders (e.g. `packetfabric_point_to_point`) so the user will not have to modify the data of one resource via a second resource? If it is impossible to implement, then we should probably throw a custom error when the user changes the subscription terms in these resources e.g. "Unable to modify subscription terms in packetfabric_point_to_point please use packetfabric_billing_modify_order instead!".

This was my original idea but we can not do it because the resources already have an UPDATE operation.

I agree with adding the warning though, good idea! @pontovinte could you add the suggested in following resources?

- Backbone VC: can only update subscription_term and/or speed
resource packetfabric_backbone_virtual_circuit

- Point-to-Point: can only update subscription_term
resource packetfabric_point_to_point

- Hosted Cloud (AWS, Azure, Google, IBM, Oracle): can only update speed
resource packetfabric_cs_<aws/azure/google/oracle/ibm>_hosted_connection

- Dedicated Cloud (AWS, Azure, Google): can only update only subscription_term and/or service_class
resource packetfabric_cs_<aws/azure/google>_dedicated_connection

- Cloud Router (AWS, Azure, Google, Oracle, IBM, IPsec, Port): can only update speed
resource packetfabric_cloud_router_connection_<aws/azure/google/oracle/ibm/ipsec/port>

**Comment by pontovinte** [comment]
> 

I'm not sure if I'm following the idea here. From billing resources, how do we identify the type the provided `circuit_id`, is there any API to call for that?
Also, I was expecting this validation to happen on server side, where does it happen today? On Portal UI?
Does the server throws any exception?

**Comment by rjouhann** [comment]
> I'm not sure if I'm following the idea here. From billing resources, how do we identify the type the provided `circuit_id`, is there any API to call for that? Also, I was expecting this validation to happen on server side, where does it happen today? On Portal UI? Does the server throw any exception?

The changes requested are not in the billing resource but in each of the resources listed above in their UPDATE logic.

Essentially, check if the user try to change any of the attributes related to billing and throw an error suggested in each of the resources mentioned below. I think it makes sense to error out here.

Example in `cloud_router_connection.go`
```
func (c *PFClient) UpdateCloudRouterConnection()
if speed is modified:
Show "Unable to modify speed, please use packetfabric_billing_modify_order resource instead!"
```

----
- **resource packetfabric_backbone_virtual_circuit**: can only update subscription_term and/or speed

If subscription_term and/or speed is modified:
Show "Unable to modify speed or subscription_term, please use packetfabric_billing_modify_order resource instead!"

- **resource packetfabric_point_to_point**: can only update subscription_term

If subscription_term is modified:
Show "Unable to modify subscription_term, please use packetfabric_billing_modify_order resource instead!"

- **resource packetfabric_cs_<aws/azure/google/oracle/ibm>_hosted_connection**: can only update speed

If speed is modified:
Show "Unable to modify speed, please use packetfabric_billing_modify_order resource instead!"

- **resource packetfabric_cs_<aws/azure/google>_dedicated_connection**: can only update only subscription_term and/or service_class

If subscription_term and/or service_class is modified:
Show "Unable to modify service_class or subscription_term, please use packetfabric_billing_modify_order resource instead!"

- **resource packetfabric_cloud_router_connection_<aws/google/ibm/ipsec/port>**: can only update speed

If speed is modified:
Show "Unable to modify speed, please use packetfabric_billing_modify_order resource instead!"

**Comment by rjouhann** [comment]
Moved the enhancement adding the warning in a different story for future. https://github.com/PacketFabric/terraform-provider-packetfabric/issues/296

**Review by caitlinwheeless — APPROVED** [review]
Checked the formatting in the doc preview tool, so everything should be ok.

**Review by rjouhann — APPROVED** [review]

**Pull request merged**